Use generator to set CMAKE_EXPORT_COMPILE_COMMANDS

donny-dont opened the pull request:
#### f6b1fa7647c7e2b2f4eefbb45c79dbd27d5479ab
<pre>
Use generator to set CMAKE_EXPORT_COMPILE_COMMANDS
<a href="https://bugs.webkit.org/show_bug.cgi?id=260666">https://bugs.webkit.org/show_bug.cgi?id=260666</a>

Reviewed by Michael Catanzaro.

In 266947@main a target `RewriteCompileCommands` was added for unified
builds where `CMAKE_EXPORT_COMPILE_COMMANDS` is `ON`. However this
is only valid for Makefile and Ninja generators. On Visual Studio it
is ignored so the target will always fail to build which was found out
because its always set to `ON` within `build-webkit`.

Move the logic for turning it `ON` into the CMake code and base it on
the `CMAKE_GENERATOR`.

* Source/cmake/WebKitCommon.cmake
* Tools/Scripts/webkitdirs.pm

Canonical link: <a href="https://commits.webkit.org/267521@main">https://commits.webkit.org/267521@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc3bf6c0e14dda65e1854e79ad03776ee9a33199

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16610 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16933 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17378 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18392 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15578 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16800 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20163 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17074 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17903 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16808 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17210 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14377 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19175 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14455 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15061 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21840 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/14343 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15442 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15227 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19530 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/15847 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15823 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13438 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/18168 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15019 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/4261 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4038 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19389 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/19388 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15648 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/4096 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/cmake/WebKitCommon.cmake:
```diff
@@ -13,6 +13,12 @@ if (NOT HAS_RUN_WEBKIT_COMMON)
+    # See https://cmake.org/cmake/help/latest/variable/CMAKE_EXPORT_COMPILE_COMMANDS.html
+    if (CMAKE_GENERATOR MATCHES "Makefile" OR CMAKE_GENERATOR MATCHES "Ninja")
+        set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+    endif ()
```
I don't like this very much. The developer or script using CMake should decide to change defaults like this. There's no value in having this enabled outside developer builds, for example.

I guess if you really want to do it here, you could gate it behind a DEVELOPER_MODE check.

donny-dont replied:
Gated behind `DEVELOPER_MODE`

donny-dont committed:
Use generator to set CMAKE_EXPORT_COMPILE_COMMANDS
https://bugs.webkit.org/show_bug.cgi?id=260666

Reviewed by Michael Catanzaro.

In 266947@main a target `RewriteCompileCommands` was added for unified
builds where `CMAKE_EXPORT_COMPILE_COMMANDS` is `ON`. However this
is only valid for Makefile and Ninja generators. On Visual Studio it
is ignored so the target will always fail to build which was found out
because its always set to `ON` within `build-webkit`.

Move the logic for turning it `ON` into the CMake code and base it on
the `CMAKE_GENERATOR`.

* Source/cmake/WebKitCommon.cmake
* Tools/Scripts/webkitdirs.pm

Canonical link: https://commits.webkit.org/267521@main
